tools/schemapi/schemapi.py

```diff
@@ -334,6 +384,109 @@ def _get_leaves_of_error_tree(
     return leaves
 
 
+def _lazy_group_tree_leaves(
+    errors: Iterable[ValidationError], /
+) -> Iterator[tuple[str, ValidationError]]:
+    """
+    Combines 3 previously distinct steps:
+
+    - ``_get_leaves_of_error_tree``
+    - (part of) ``_group_errors_by_json_path``
+        - Doesnt actually group yet, can by calling `dict(result)`.
+    - ``_is_required_value_error``
+    """  # noqa: D400
+    for err in errors:
+        if err_context := err.context:
+            yield from _lazy_group_tree_leaves(err_context)
+        elif err.validator == "required" and err.validator_value == ["value"]:
+            continue
+        else:
+            yield _json_path(err), err
+
+
+_fn_path = cast("Callable[[tuple[str, ValidationError]], str]", operator.itemgetter(0))
+"""Key function for ``(json_path, ValidationError)``."""
+_fn_validator = cast("Callable[[ValidationError], str]", operator.attrgetter("validator"))  # fmt: off
+"""Key function for ``ValidationError.validator``."""
+
+
+def _lazy_subset_to_most_specific_json_paths(
+    json_path_errors: Iterator[tuple[str, ValidationError]], /
+) -> Iterator[Iterable[ValidationError]]:
+    """
+    Currently using a `list`, but typing it more restrictive to see if it can be avoided.
+
+    - Needs to be sorted to work with groupby
+    - Reversing allows prioritising more specific groups, since they are seen first
+    - Then re-reversed, to keep seen order
+
+    """
+    rev_sort = sorted(json_path_errors, key=_fn_path, reverse=True)
+    keeping: dict[str, Iterable[ValidationError]] = {}
+    for unique_path, grouped_errors in groupby(rev_sort, key=_fn_path):
+        if any(seen.startswith(unique_path) for seen in keeping):
+            continue
+        else:
+            keeping[unique_path] = [err for _, err in grouped_errors]
+    yield from reversed(keeping.values())
```

Comment:
TODO: Test rewriting this as:

```py
yield from islice(reversed(keeping.values()), 3)
```

Thinking it would skip some work that is discarded in `SchemaValidationError._get_message()`.

Reply:
Fixed in [`bbf6f14` (#3547)](https://github.com/vega/altair/pull/3547/commits/bbf6f142ac6a4df0dc6d7e3ca326b9a8d32b413e)